From 2fbb62780614245a2e8862dae05c527340c8c813 Mon Sep 17 00:00:00 2001
From: Shawn Lin <shawn.lin@rock-chips.com>
Date: Tue, 15 Dec 2020 14:35:57 +0800
Subject: [PATCH] PCI: rockchip: dw: Add loopbask master support

In order to make loopback slave into LPBK.Active state,
we add "rockchip,lpbk-master" for DTU to force it as a
remote loopback master device. Nothing needed for slave.

Change-Id: Id851ad8e34e29939f12f76cb132d942bc059a811
Signed-off-by: Shawn Lin <shawn.lin@rock-chips.com>
---
 drivers/pci/controller/dwc/pcie-designware.h  | 1 +
 drivers/pci/controller/dwc/pcie-dw-rockchip.c | 7 +++++++
 2 files changed, 8 insertions(+)

diff --git a/drivers/pci/controller/dwc/pcie-designware.h b/drivers/pci/controller/dwc/pcie-designware.h
index 2d7d64fdff76..d125e36b01d1 100644
--- a/drivers/pci/controller/dwc/pcie-designware.h
+++ b/drivers/pci/controller/dwc/pcie-designware.h
@@ -36,6 +36,7 @@
 #define PORT_LINK_MODE_2_LANES		(0x3 << 16)
 #define PORT_LINK_MODE_4_LANES		(0x7 << 16)
 #define PORT_LINK_MODE_8_LANES		(0xf << 16)
+#define PORT_LINK_LPBK_ENABLE		(0x1 << 2)
 
 #define PCIE_TIMER_CTRL_MAX_FUNC_NUM	0x718
 #define FAST_LINK_SCALING_FACTOR	0x9fffffff
diff --git a/drivers/pci/controller/dwc/pcie-dw-rockchip.c b/drivers/pci/controller/dwc/pcie-dw-rockchip.c
index ce390f415ba4..92e5788ca568 100644
--- a/drivers/pci/controller/dwc/pcie-dw-rockchip.c
+++ b/drivers/pci/controller/dwc/pcie-dw-rockchip.c
@@ -1116,6 +1116,7 @@ static int rk_pcie_probe(struct platform_device *pdev)
 	const struct rk_pcie_of_data *data;
 	enum rk_pcie_device_mode mode;
 	struct device_node *np = pdev->dev.of_node;
+	u32 val;
 
 	match = of_match_device(rk_pcie_of_match, dev);
 	if (!match)
@@ -1199,6 +1200,12 @@ static int rk_pcie_probe(struct platform_device *pdev)
 	/* Set PCIe mode */
 	rk_pcie_set_mode(rk_pcie);
 
+	if (device_property_read_bool(dev, "rockchip,lpbk-master")) {
+		val = dw_pcie_readl_dbi(pci, PCIE_PORT_LINK_CONTROL);
+		val |= PORT_LINK_LPBK_ENABLE;
+		dw_pcie_writel_dbi(pci, PCIE_PORT_LINK_CONTROL, val);
+	}
+
 	switch (rk_pcie->mode) {
 	case RK_PCIE_RC_TYPE:
 		ret = rk_add_pcie_port(rk_pcie);
-- 
2.35.3

